typing/shape_reduce.mli

```diff
@@ -18,10 +18,10 @@
 (** The result of reducing a shape and looking for its uid *)
 type result =
   | Resolved of Shape.Uid.t (** Shape reduction succeeded and a uid was found *)
-  | Resolved_alias of Shape.Uid.t list (** Reduction led to an alias chain *)
+  | Resolved_alias of Shape.Uid.t * result (** Reduction led to an alias *)
```

Comment:
If possible I would rather avoid changing the interface for OCaml 5.2.0, but I am not sure if we are still aiming for 5.2.0 for this PR.

Reply:
> If possible I would rather avoid changing the interface for OCaml 5.2.0

Is it because of the already released alpha and beta versions with bumped magic numbers ? Since the `result` type was actually introduced in 5.2 itself, it seems like it could be better to change it right now... 

I can also easily rollback to the first iteration on this PR that fixes the performance issue without changing the type if that is required.

> I am not sure if we are still aiming for 5.2.0 for this PR

The discussion diverged to other unclear parts of the shape reduction, but I think this PR's original scope is smaller than that and mostly uncontroversial: removing extraneous work done in some cases, with an agreed-on fix (stepped read-back). We do need to re-think shape reduction and it's handling of uids at some point, but that's out-of-scope for that pr. What do you think @gasche ?

Reply:
It is more a question of API stability for shape clients. In particular, this change would require a new beta and a patch to odoc. This is still ok-ish, if the PR converges this week, but this is starting to get late in the release cycle for this kind of changes.

Reply:
I understand, thanks for the details. Now that you mention odoc, it reminds me that the performance fix might actually be important for their usage of the shapes. (We discussed it a few weeks ago with @panglesd  and @Julow, they perform similar actions as we do in Merlin to identify definitions' uids.)

Reply:
`odoc` would indeed would benefit from the improvements of this PR! (The "render source code" feature of `odoc` is probably very inefficient by itself, but we will work on it when it is usable by drivers.)

Regarding API changes, we haven't yet released a 5.2 compatible version of odoc, so I think that for us it is still fine to include it.

Reply:
Ok, let's have a beta 2 and I will amend my patch for odoc for 5.2.